#!/usr/bin/env Rscript

###############################
#    RAS Modifiers Project    #
###############################

# Copyright (c) 2023-Present Ryan L. Collins and the Van Allen/Gusev/Haigis Laboratories
# Distributed under terms of the GNU GPL v2.0 License (see LICENSE)
# Contact: Ryan L. Collins <Ryan_Collins@dfci.harvard.edu>

# Generate full-slide summary diagram of KRAS somatic mutations in dataset


#########
# Setup #
#########
# Load necessary libraries and constants
require(OncoModR, quietly=TRUE)
require(argparse, quietly=TRUE)
require(meta, quietly=TRUE)
OncoModR::load.constants("all")


##################
# Data functions #
##################
# Load & clean somatic frequency data from .tsv
load.somatic.freqs <- function(tsv.in){
  # Read data into memory
  data <- read.table(tsv.in, sep="\t", comment.char="", check.names=F, header=T)
  rownames(data) <- data$csq
  data <- data[which(data$gene == "KRAS"), ]

  # Drop unnecessary columns
  drop.cols <- c("position", "gene",
                 colnames(data)[grep("_set_id$|vids$", colnames(data))])
  data <- data[, -which(colnames(data) %in% drop.cols)]

  # Compute inverse variance-weighted mean freq for all mutations per cancer type
  cancers <- unique(unlist(sapply(strsplit(colnames(data)[grep("_AF$", colnames(data))],
                                           split="_"), function(l){l[2]})))
  for(cancer in cancers){
    mfreqs <- as.data.frame(do.call("rbind", lapply(1:nrow(data), function(i){
      n.all <- as.numeric(data[i, grep(paste("", cancer, "AN", sep="_"), colnames(data))])
      n.hit <- as.numeric(data[i, grep(paste("", cancer, "AC", sep="_"), colnames(data))])
      meta.res <- metaprop(n.hit, n.all, method="Inverse")
      meta:::backtransf(unlist(meta.res[c("TE.random", "lower.random", "upper.random")]),
                        sm=meta.res$sm)
    })))
    colnames(mfreqs) <- paste("meta_", cancer, "_AF", c("", "_lower", "_upper"), sep="")
    data <- cbind(data, mfreqs)
  }

  # Return sorted df
  data[, sort(colnames(data))]
}


######################
# Plotting functions #
######################
# Helper function to plot a single panel of variant frequency meta-analyses
plot.freq.meta.single <- function(data, csq, cancer, title=NULL, text.cex=0.75,
                                  diamond.ratio=1/4, proportional.bars=TRUE,
                                  x.space=1/10, ylims=NULL, label.y.axis=TRUE,
                                  parmar=c(0.5, 3, 1, 0.5)){
  # Get data
  cohorts <- unique(cohort.names.short)
  point.ests <- as.numeric(data[csq, paste(cohorts, cancer, "AF", sep="_")])
  sample.sizes <- as.numeric(data[csq, paste(cohorts, cancer, "AN", sep="_")])
  meta.stats <- as.numeric(data[csq, paste("meta", cancer,
                                           c("AF", "AF_lower", "AF_upper"),
                                           sep="_")])

  # Get plot dimensions
  xlims <- c(0, 1 + (2 * x.space))
  if(is.null(ylims)){
    ylims <- c(0, max(point.ests, meta.stats))
  }
  diamond.x <- x.space + (c(0, 0.5, 1, 0.5) * diamond.ratio)
  diamond.y <- meta.stats[c(1, 2, 1, 3)]
  if(proportional.bars){
    bar.xd <- c(0, cumsum(sample.sizes))/sum(sample.sizes)
  }else{
    bar.xd <- seq(0, 1, length=length(cohorts) + 1)
  }
  bar.xleft <- (2 * x.space) + diamond.ratio + ((1 - diamond.ratio) * bar.xd[-length(bar.xd)])
  bar.xright <- (2 * x.space) + diamond.ratio + ((1 - diamond.ratio) * bar.xd[-1])

  # Prep plot area
  prep.plot.area(xlims=xlims, ylims=ylims, parmar=parmar)
  if(label.y.axis){
    y.labels <- paste(100 * axTicks(2), "%", sep="")
    y.title <- "Pct. of Tumors"
  }else{
    y.labels <- rep("", length(axTicks(2)))
    y.title <- ""
  }
  clean.axis(2, infinite=TRUE, labels=y.labels, title=y.title, title.line=1.25)
  clean.axis(1, tck=0, labels=NA)
  mtext(title, side=3, font=2, cex=text.cex)

  # Add diamond & bars
  polygon(x=diamond.x, y=diamond.y, col=cancer.colors[cancer])
  segments(x0=diamond.x[1:2], x1=diamond.x[3:2],
           y0=meta.stats[1:2], y1=meta.stats[c(1, 3)],
           col="white")
  polygon(x=diamond.x, y=diamond.y, border="black", col=NA)
  text(x=diamond.x[3], y=diamond.y[4]-(0.02*diff(par("usr")[3:4])),
       pos=3, cex=text.cex, xpd=T,
       labels=paste(round(100 * meta.stats[1], 1), "%", sep=""))
  rect(xleft=bar.xleft, xright=bar.xright, ybottom=0, ytop=point.ests,
       col=cancer.palettes[[cancer]][cohort.color.prefixes[cohorts]])
}


###########
# RScript #
###########
# Parse command line arguments and options
parser <- ArgumentParser(description="Summarize all KRAS somatic mutation data")
parser$add_argument("--stats", metavar=".tsv", type="character", required=TRUE,
                    help=paste("somatic mutation frequency .tsv generated by",
                               "gather_somatic_ras_data.py"))
parser$add_argument('--out-pdf', metavar='path', type="character", nargs=1,
                    help='output .pdf')
args <- parser$parse_args()

# # DEV
# args <- list("stats"="~/scratch/ras_somatic_variants.tsv.gz",
#              "out_pdf"="~/scratch/KRAS_summary_plot.pdf")

# Load somatic mutation frequency data
data <- load.somatic.freqs(args$stats)
data$csq.simple <- paste(data$ref, data$codon, data$alt, sep="")

# Restrict frequency data to top N non-CNA mutations
n.mut.highlights <- 10
freq.order <- order(apply(data[, grep("meta_.*_AF$", colnames(data))], 1, max), decreasing=TRUE)
highlight.muts <- head(setdiff(rownames(data)[freq.order], c("AMP", "DEL")), n.mut.highlights)
top.data <- data[which(rownames(data) %in% highlight.muts), ]
top.data <- top.data[with(top.data, order(codon, alt)), ]

# Set hard-coded plotting variables
cancer.order <- c("PDAC", "CRAD", "LUAD")
ctype.ylims <- lapply(cancer.order, function(cancer){
  c(0, 1.25*max(top.data[, grep(paste(cancer, "AF", sep="_"), colnames(top.data))]))
})
names(ctype.ylims) <- cancer.order

# Plot main figure
pdf(args$out_pdf, width=180/25.4, height=2.5)
layout(matrix(c(rep(1, n.mut.highlights), (1:(3*n.mut.highlights))+1),
              nrow=4, byrow=T), heights=c(0.5, 1, 1, 1))
par(mar=rep(0, 4)); plot(1)
for(k in 1:length(cancer.order)){
  for(i in 1:n.mut.highlights){
    plot.freq.meta.single(top.data, rownames(top.data)[i], cancer=cancer.order[k],
                          ylims=ctype.ylims[[k]], label.y.axis=FALSE,
                          title=if(k==1){top.data[i, "csq.simple"]}else{NULL},
                          parmar=c(0.5, 1, 1, 0.5))
  }
}
dev.off()
